nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py

```diff
@@ -34,11 +75,57 @@ def __init__(self, dim: int, seq_len_interpolation_factor: int = None):
             by this factor via the trick in https://arxiv.org/abs/2306.15595.
         """
         super().__init__()
+        self.use_yarn = use_yarn
+        self.base = base
         self.seq_len_interpolation_factor = seq_len_interpolation_factor
-        inv_freq = 1.0 / (10000 ** (torch.arange(0, dim, 2).float() / dim))
-        self.register_buffer('inv_freq', inv_freq)
+        if self.use_yarn:
+            self.seq_len_interpolation_factor = seq_len_interpolation_factor  # scale
+            self.pretrain_max_positional_embeddings = pretrain_max_positional_embeddings
+            self.extrapolation_factor = extrapolation_factor
+            self.attn_factor = attn_factor
+            self.beta_fast = beta_fast
+            self.beta_slow = beta_slow
+            self.dim = dim
+            pos_freqs = self.base ** (torch.arange(0, self.dim, 2).float() / self.dim)
+            inv_freq_extrapolation = 1.0 / pos_freqs
+            inv_freq_interpolation = 1.0 / (self.seq_len_interpolation_factor * pos_freqs)
+
+            low, high = find_correction_range(
+                self.beta_fast, self.beta_slow, self.dim, self.base, self.pretrain_max_positional_embeddings
+            )
+            inv_freq_mask = (
+                1 - linear_ramp_mask(low, high, self.dim // 2).float()
+            ) * self.extrapolation_factor  # Get n-d rotational scaling corrected for extrapolation
+            inv_freq = inv_freq_interpolation * (1 - inv_freq_mask) + inv_freq_extrapolation * inv_freq_mask
+
+            self.register_buffer("inv_freq", inv_freq)
+            self.mscale = float(
+                get_mscale(self.seq_len_interpolation_factor) * self.attn_factor
+            )  # Get n-d magnitude scaling corrected for interpolation
+
+            self.max_seq_len_cached = max_position_embeddings
```

Comment:
If we already have `seq_len_interpolation_factor`, do we still need `max_position_embeddings`?

Reply:
probably not, I'll remove it and replace usages of `max_position_embeddings` with `seq_len_interpolation_factor * pretrain_max_positional_embeddings`